### PR TITLE
Make dynamic shared informer stoppable

### PR DESCRIPTION
**Pull request by yliaog**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
 /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
when CRD is removed, the dynamic shared informer for it needs to be able to stop.

currently (before the fix), when CRD is removed, then the same CRD is recreated, the two CRDs would share the same Informer, which would not work.

after the fix, when CRD is removed, the informer is stopped and removed from the factory, when the same CRD is recreated, a new Informer would be created for it.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
None
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer_test.go [hunk 1]
```diff
@@ -131,15 +155,29 @@ func TestDynamicSharedInformerFactory(t *testing.T) {
...
+				close(stop)
+			}
 
 			testObject := ts.trigger(ts.gvr, ts.ns, fakeClient, ts.existingObj)
 			select {
 			case objFromInformer := <-informerReciveObjectCh:
-				if !equality.Semantic.DeepEqual(testObject, objFromInformer) {
-					t.Fatalf("%v", diff.ObjectDiff(testObject, objFromInformer))
+				if ts.isStop {
+					if equality.Semantic.DeepEqual(testObject, objFromInformer) {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer_test.go]: I think no matter what object is received from the channel, it's an error.

File: staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go [hunk 2]
```diff
@@ -138,20 +153,79 @@ func NewFilteredDynamicInformer(client dynamic.Interface, gvr schema.GroupVersio
...
+	return d
 }
 
 func (d *dynamicInformer) Lister() cache.GenericLister {
 	return dynamiclister.NewRuntimeObjectShim(dynamiclister.New(d.informer.GetIndexer(), d.gvr))
 }
+
+var _ cache.SharedIndexInformer = &dynamicInformer{}
+
+// It would be much cleaner if we could embed the struct cache.sharedIndexInformer in struct dynamicInformer
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go]: Why not embedding cache.SharedIndexInformer?

File: staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go [hunk 3]
```diff
@@ -138,20 +153,79 @@ func NewFilteredDynamicInformer(client dynamic.Interface, gvr schema.GroupVersio
...
+func (d *dynamicInformer) GetIndexer() cache.Indexer {
+	return d.informer.GetIndexer()
+}
+
+type dummyController struct {
+	informer *dynamicInformer
+}
+
+func (v *dummyController) Run(stopCh <-chan struct{}) {
+	<-stopCh
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go]: Can we just make the caller call factory.RemoveResource() directly, and let RemoveResource() stop the informer. I think it's better than letting caller manage individual stop channels for every resource.

**Review by caesarxuchao — COMMENTED** [review]
I assume this pull request is fixing the gc bug. Could you add more details to the pull request description?

Please take my suggestions with a grain of salt. I don't think I understand the pull request entirely.

**Comment by caesarxuchao** [comment]
> currently (before the fix), when CRD is removed, then the same CRD is recreated, the two CRDs would share the same Informer, which would not work.

Even with this fix, it's possible that the caller couldn't distinguish the recreated CRD from the old CRD, thus failing to stop the informer.

In GC's case, the GC re-discovers every [30 seconds](https://github.com/kubernetes/kubernetes/blob/851afa0beaac20757ad2556af94e31dce0a918be/cmd/kube-controller-manager/app/core.go#L416). If a CRD is deleted and then recreated within 30s, GC could miss both the deletion and creation, then this fix couldn't prevent the GC bug, right?

File: staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go (hunk 2 above)
Comment (yliaog) [reply staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go]: switched to embed the interface cache.SharedIndexInformer

File: staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go [hunk 4]
```diff
@@ -138,20 +152,65 @@ func NewFilteredDynamicInformer(client dynamic.Interface, gvr schema.GroupVersio
...
+var _ cache.SharedIndexInformer = &dynamicInformer{}
+
+func (d *dynamicInformer) GetController() cache.Controller {
+	return &dummyController{informer: d}
+}
+
+// Use local stopCh, the passed in stopCh is ignored.
+// The informer is stopped via its controller
+func (d *dynamicInformer) Run(stopCh <-chan struct{}) {
+	d.SharedIndexInformer.Run(d.stopCh)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go]: d.Run will be called multiple times as it's a shared dynamic informer. Everytime, it calls to d.SharedIndexInformer.Run, which will [create](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L261) a new concrete cache.controller with reflector and cache. This defeats the purpose of a shared informer.

**Review by caesarxuchao — COMMENTED** [review]
I also have some high-level questions on the use case.

Are the garbage collector and the resource quota controller going to share the same dynamic informer factory? (I assume the answer is yes)

Does the gc and resource quota controller share the same discovery resync loop, or do they do the resync separately? (I assume it's the latter). What if there is race between the resync loops? i.e., one of them just creates a new cache.sharedIndexInformer, and the other resync loop deletes sharedIndexInformer?

**Comment by yliaog** [comment]
for the resync loop races:

1: t1) GC creates informer#1 t2) RQ gets informer#1 t3) GC deletes informer#1 t4) GC creates informer#2 t5) RQ deletes informer#1 t6) RQ gets infomer#2

this scenario, it's key to ensure RQ deletes informer#1, not the informer#2 that is just created by GC. the factory removeResource will check if the informer to remove is the informer that is stopped, if not, removeResource will skip.

2: t1) GC creates informer#1 t2) RQ gets informer#1 t3) GC deletes informer#1 t4) RQ deletes informer#1 t5) GC creates informer#2  t6) RQ gets infomer#2

this scenario should be fine.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/78029 stops watches when CRDs change

**Comment by yliaog** [comment]
PR #78029 merged, close this PR.

/close

**Pull request closed without merging**